pkg/didcomm/protocol/didexchange/service.go

```diff
@@ -86,6 +150,92 @@ func (s *Service) Handle(msg dispatcher.DIDCommMsg) error {
 	return nil
 }
 
+func (s *Service) sendEvent(msg dispatcher.DIDCommMsg, threadID string, nextState state) (bool, error) {
+	invoked := false
+
+	// invoke events for ConnectionRequest, ConnectionResponse or ConnectionAck
+	if msg.Type != ConnectionRequest &&
+		msg.Type != ConnectionResponse && msg.Type != ConnectionAck {
+		return false, nil
+	}
+
+	jsonDoc, err := json.Marshal(&message{
+		Msg:           msg,
+		ThreadID:      threadID,
+		NextStateName: nextState.Name(),
+	})
+
+	if err != nil {
+		return false, fmt.Errorf("JSON marshalling of document failed: %w", err)
+	}
+
+	// save the incoming message in the store (to retrieve later when callback events are fired)
+	id := s.generateRandomID()
+	err = s.store.Put(id, jsonDoc)
+	if err != nil {
+		return false, fmt.Errorf("JSON marshalling of document failed: %w", err)
+	}
+
+	// create the message for the channel
+	didCommEvent := event.DIDCommEvent{
+		Message: msg,
+		Callback: func(didCommCallback event.DIDCommCallback) {
+			s.processCallback(id, didCommCallback)
+		},
+	}
+
+	// trigger the registered go channels
+	for _, handler := range s.event {
+		handler <- didCommEvent
+		invoked = true
+	}
+
+	return invoked, nil
+}
+
+// startInternalListener listens to messages in gochannel for callback messages from clients.
+func (s *Service) startInternalListener() {
+	go func() {
+		for msg := range s.ch {
+			// TODO handle error in callback - https://github.com/hyperledger/aries-framework-go/issues/242
```

Comment:
Also need to ensure we have tests for:

- no listeners call the callback.
- multiple listeners call the callback.
- some listeners call the callback and some do not.